### PR TITLE
[W5.4][T14-A4]Le Quang Quan

### PR DESCRIPTION
**Pull request by quanle1994**
Add edit command into the Address Book

### CODE REVIEW

File: src/seedu/addressbook/Main.java [hunk 1]
```diff
@@ -3,11 +3,20 @@
 import javafx.application.Application;
 import javafx.application.Platform;
 
+import javafx.fxml.FXMLLoader;
```
Comment (ang-zy) [inline src/seedu/addressbook/Main.java:6]: Please keep your PRs separate. Your PR should only have changes pertaining to your PR title.

File: test/java/seedu/addressbook/logic/LogicTest.java [hunk 2]
```diff
@@ -200,12 +206,137 @@ public void execute_list_showsAllPersons() throws Exception {
                               true,
                               expectedList);
     }
+    @Test
```
Comment (ang-zy) [inline test/java/seedu/addressbook/logic/LogicTest.java:209]: Would it be more appropriate if these tests were separated into their own file?

File: src/seedu/addressbook/parser/Parser.java [hunk 3]
```diff
@@ -116,7 +116,8 @@ public Command parseCommand(String userInput) {
                 return prepareDelete(arguments);
 
             case EditCommand.COMMAND_WORD:
-                return processEdit(arguments);
+//                return processEdit(arguments);
```
Comment (ang-zy) [inline src/seedu/addressbook/parser/Parser.java]: Don't comment out unused code. Just delete it. You can restore it by reversing/referencing the commit.

File: src/seedu/addressbook/commands/EditCommand.java [hunk 4]
```diff
@@ -0,0 +1,58 @@
...
+     * @param name name of the person
+     * @param phone phone number of the person
+     * @param isPhonePrivate private status of phone number
+     * @param email email of the person
+     * @param isEmailPrivate private status of email
+     * @param address address of the person
+     * @param isAddressPrivate private status of address
+     * @param tagArguments tags to the person
+     */
+    public EditCommand(String index, String name,
```
Comment (ang-zy) [inline src/seedu/addressbook/commands/EditCommand.java]: What is the purpose of this class?

File: doc/UserGuide.md [hunk 5]
```diff
@@ -34,6 +34,32 @@ Examples:
...
+> items with `...` after them can have multiple instances. Order of parameters are fixed. 
+> 
+> Put a `p` before the phone / email / address prefixes to mark it as `private`. `private` details can only
+> be seen using the `viewall` command.
+> 
+> Persons can have any number of tags (including 0)
+
+Examples:
+* `edit`
+* `edit n/John Doe p/98765432 e/johnd@gmail.com a/John street, block 123, #01-01`
```
Comment (ang-zy) [inline doc/UserGuide.md]: Where's the index in the examples?

File: src/seedu/addressbook/parser/Parser.java [hunk 6]
```diff
@@ -226,5 +289,41 @@ private Command prepareFind(String args) {
...
+        if (!matcher.matches()) {
+            return new IncorrectCommand(String.format(MESSAGE_INVALID_COMMAND_FORMAT, EditCommand.MESSAGE_USAGE));
+        }
+        try {
+            Set<String> tags = getTagsFromArgs(matcher.group("tagArguments"));
+            final Set<Tag> tagSet = new HashSet<>();
+            for (String tagName : tags) {
+                tagSet.add(new Tag(tagName));
+            }
+            return new DeleteCommand(Integer.parseInt(matcher.group("index")), true);
```
Comment (ang-zy) [inline src/seedu/addressbook/parser/Parser.java]: Does this really edit the person?

File: src/seedu/addressbook/commands/EditCommand.java [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+                       String phone, boolean isPhonePrivate,
+                       String email, boolean isEmailPrivate,
+                       String address, boolean isAddressPrivate,
+                       Set<String> tagArguments){
+    }
+
+    @Override
+    public CommandResult execute() {
+        return new CommandResult(MESSAGE_SUCCESS);
```
Comment (ang-zy) [inline src/seedu/addressbook/commands/EditCommand.java]: This might give you false positives for your JUnit Test

**Review by ang-zy — CHANGES_REQUESTED** [review]
Your implementation of edit is extremely convoluted. Try to keep as much of the edit functionality within the EditCommand class as possible.
Good effort on the JUnit tests and the user guide, but I can't give you the marks because your functionality doesn't work properly.

File: doc/UserGuide.md (hunk 5 above)
Comment (quanle1994) [reply doc/UserGuide.md]: Oh yah I forgot the index, the command should look like this. Can I edit the code now?
edit 1 n/John Doe p/98765432 e/johnd@gmail.com a/John street, block 123, #01-01

File: src/seedu/addressbook/commands/EditCommand.java (hunk 4 above)
Comment (quanle1994) [reply src/seedu/addressbook/commands/EditCommand.java]: Yup I should have removed this class.

**Comment by quanle1994** [comment]
Hi zy. Thank you for reviewing my code. I was wondering why my code did not function properly. Was it because in my help function I missed out the index? The logic is that when you edit a person, the system will delete the existing person and add in a new entry, so my edit command is a combination of delete and add commands. Could I still get the marks if I still use that logic and update the changes that you want? Coz I thought by doing that I did edit the person in the system.

**Comment by ang-zy** [comment]
Your idea for the edit command can work, but it is very poor software engineering as your edit command will rely on the functionality of delete and add. If those 2 commands change at any point of time, your edit implementation may have to change as well.
Also, when I traced your edit command code, I didn't see anywhere where the person was added.
I can award you the marks if you can update the PR such that the edit functionality works, and your tests and documentations are updated by the end of recess week

**Comment by quanle1994** [comment]
Hi Zhi Yuan. Thank you for extending the deadline for me. I agree that this currently logic is very dependent on the other 2 commands. I have updated the user guide and the help return message to the correct command. The logic is still the same as what I have stated before. I will also at the same time attempt to do a proper edit command implementation by Friday but in the meantime, could you grade my work based on what I have currently?

**Comment by ang-zy** [comment]
Unfortunately the grading is binary. You need to have everything (i.e. working functionality, JUnit test and documentation) before I can give you the marks.

**Comment by quanle1994** [comment]
What I meant is that my edit function already works in the current pull request, but it works in the way I described above (delete the person by the index indicated, then add a new person with the particulars given in the command arguments). So if you follow the commands I have updated in the user guide and the help return message, it should do what I want it to do as per stated, so I'm just wondering if this is sufficient to get the marks because another way of implementing this command is to mutate the addressBook instance then write it to the storage file and I didn't do this.

**Comment by ang-zy** [comment]
I just pulled your fork, but I can't get it working. Please show me the functionality during tutorial on Thursday.

File: src/seedu/addressbook/commands/EditCommand.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+                new Email(email, isEmailPrivate),
+                new Address(address, isAddressPrivate),
+                new UniqueTagList(tagSet)
+        );
+    }
+
+    @Override
+    public CommandResult execute() {
+        try {
+            final ReadOnlyPerson target = getTargetPerson();
```
Comment (ang-zy) [inline src/seedu/addressbook/commands/EditCommand.java:53]: Is this variable used?

**Review by ang-zy — COMMENTED** [review]
This is a much better implementation than the first one! It doesn't rely on the other functionality of the other 2 commands.

**Comment by ang-zy** [comment]
Close the PR after you're done reading the comments

File: src/seedu/addressbook/commands/EditCommand.java (hunk 8 above)
Comment (quanle1994) [reply src/seedu/addressbook/commands/EditCommand.java:53]: Oh I'm not gonna use the target variable. I just want to call the getTargetPerson() method for the programme to update the listing.

**Pull request closed without merging**